File: java/547.java
class Solution {
    private void ma(int ic[][], int i, int n, boolean trav[]) {
        if (trav[i])
            return;
        trav[i] = true;
        for (int j = 0; j < n; j++)
            if (ic[i][j] == 1)
                ma(ic, j, n, trav);
    }

    public int findCircleNum(int[][] ic) {
        int n = ic.length, prov = 0;
        boolean trav[] = new boolean[n];
        for (int i = 0; i < n; i++)
            if (!trav[i]) {
                ma(ic, i, n, trav);
                prov++;
            }
        return prov;
    }
}

File: python3/547.py
class Solution:
    def findCircleNum(self, ic: List[List[int]]) -> int:
        def ma(ic, i, n, trav):
            if trav[i]:
                return
            trav[i] = True
            for j in range(n):
                if ic[i][j] == 1:
                    ma(ic, j, n, trav)

        n = len(ic)
        prov = 0
        trav = [False] * n
        for i in range(n):
            if not trav[i]:
                ma(ic, i, n, trav)
                prov += 1
        return prov
